bookshelf/src/CloudsqlModel.php

```diff
@@ -0,0 +1,163 @@
+        $this->columnNames = array_map(function ($columnDefinition) {
+            return explode(' ', $columnDefinition)[0];
+        }, $columns);
+        $columnText = implode(', ', $columns);
+        $pdo = $this->newConnection();
+        $pdo->query("CREATE TABLE IF NOT EXISTS books ($columnText)");
+    }
+
+    public function listBooks($limit = 10, $cursor = null)
+    {
+        $pdo = $this->newConnection();
+        if ($cursor) {
+            $query = 'SELECT * FROM books WHERE id > :cursor ORDER BY id'.
+                ' LIMIT :limit';
+            $statement = $pdo->prepare($query);
+            $statement->bindValue(':cursor', $cursor, PDO::PARAM_INT);
+        } else {
+            $query = 'SELECT * FROM books ORDER BY id LIMIT :limit';
+            $statement = $pdo->prepare($query);
+        }
+        $statement->bindValue(':limit', $limit + 1, PDO::PARAM_INT);
+        $statement->execute();
+        $rows = array();
+        $last_row = null;
+        $new_cursor = null;
+        while ($row = $statement->fetch(PDO::FETCH_ASSOC)) {
+            if (count($rows) == $limit) {
+                $new_cursor = $last_row['id'];
+                break;
+            }
+            array_push($rows, $row);
+            $last_row = $row;
+        }
+
+        return array(
+            'books' => $rows,
+            'cursor' => $new_cursor,
+        );
+    }
+
+    public function create($book, $id = null)
+    {
+        $names = '';
+        $values = [];
+        $question_marks = '';
+        $separator = '';
+        $pdo = $this->newConnection();
+        foreach ($this->columnNames as $column) {
+            if (array_key_exists($column, $book)) {
+                $question_marks = "$question_marks$separator?";
+                $names = "$names$separator$column";
+                $separator = ', ';
+                array_push($values, $book[$column]);
+            }
+        }
+        $statement = $pdo->prepare(
+            "INSERT INTO books ($names) VALUES($question_marks)"
+        );
+        $statement->execute($values);
+
+        return $pdo->lastInsertId();
+    }
+
+    public function read($id)
+    {
+        $pdo = $this->newConnection();
+        $statement = $pdo->prepare('SELECT * FROM books WHERE id = :id');
+        $statement->bindValue('id', $id, PDO::PARAM_INT);
+        $statement->execute();
+
+        return $statement->fetch(PDO::FETCH_ASSOC);
+    }
+
+    public function update($book)
+    {
+        $assignments = '';
+        $values = [];
+        $separator = '';
+        $pdo = $this->newConnection();
+        foreach ($this->columnNames as $column) {
+            if (array_key_exists($column, $book)) {
+                $assignments = "$assignments$separator$column=?";
+                $separator = ', ';
+                array_push($values, $book[$column]);
+            }
+        }
+        $statement = $pdo->prepare(
+            "UPDATE books SET $assignments WHERE id = ?"
+        );
```

Comment:
Same thing here as in `create`, I would like to see `$values[] = ...` instead of `array_push`, and `implode` or `sprintf` instead of embedding variables inside double quotes. Even better yet would be to avoid the `foreach`.